### ISSUE TITLE
Kubernetes - SERVFAIL if we cycle our apiserver certs

### ISSUE THREAD

**Issue by mmack** [coredns/coredns#2464]
We cycle our apiserver certs on a regular basis. In the same time coredns is not able to answer dns queries. 

Hard facts:
```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.2", GitCommit:"cff46ab41ff0bb44d8584413b598ad8360ec1def", GitTreeState:"clean", BuildDate:"2019-01-10T23:35:51Z", GoVersion:"go1.11.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.5", GitCommit:"753b2dbc622f5cc417845f0ff8a77f539a4213ea", GitTreeState:"clean", BuildDate:"2018-11-26T14:31:35Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
``` 
```
.:53
2019-01-14T21:20:42.732Z [INFO] CoreDNS-1.3.1
2019-01-14T21:20:42.732Z [INFO] linux/amd64, go1.11.4, 6b56a9c
CoreDNS-1.3.1
linux/amd64, go1.11.4, 6b56a9c
```

Config:
```
    .:53 {
        errors
        health
        autopath @kubernetes
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods verified
          upstream
          fallthrough in-addr.arpa ip6.arpa
          ttl 5
        }
        prometheus :9153
        proxy . /etc/resolv.conf {
          policy sequential
        }
        cache 300
        reload
    }
```

Below a screenshot of a visualization of logs. The green bars are apiserver restarts. Other colors are several dns queries failing. Last 24 hours. The 2 reloads to the right are reloads of the standby master 2. The reload of the active master is barely visible between the dns errors.
 
![screenshot from 2019-01-15 08-47-12](https://user-images.githubusercontent.com/100767/51168195-d80fd780-18a8-11e9-90a6-6e1cbb72541f.png)

There are two other guys in the slack channel with the same problem. Maybe they can post their findings here...

Greetings,
Max

**Comment by chrisohaver** [comment]
Can you point to a resource how to set up cert cycling?  I want to better understand the process, and perhaps replicate locally to test various fixes.

**Comment by mmack** [comment]
We're running kubelet as a systemd unit. The kubelet ssl config looks like this:

```
--client-ca-file=/etc/kubernetes/ssl/kubelet-ca.pem
--tls-cert-file=/etc/kubernetes/ssl/kubelet.pem           
--tls-private-key-file=/etc/kubernetes/ssl/kubelet-key.pem
``` 

Then some process creates new ssl certs and this unit kicks in:

```
[Unit]
Description=restarts all k8s components after some file inside /etc/kubernetes/ssl changed
After=network.target

[Service]
Type=oneshot
ExecStart=/usr/bin/systemctl restart apiserver controller scheduler kubelet cert-server

[Install]
WantedBy=multi-user.target
```

**Comment by ffilippopoulos** [comment]
We are observing the same behaviour. We currently run kube 1.12.3 (we have clusters both on aws and gcp) and we are restarting kubelets daily to rotate certs (includes restarting `kubelet` systemd unit).

Our coredns conf looks like:
```
    .:53 {
        errors
        health
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          endpoint_pod_names
          upstream
          fallthrough in-addr.arpa ip6.arpa
          ttl 5
        }

        prometheus :9153
        proxy . /etc/resolv.conf
        cache 30
        loop
        reload
        loadbalance
    }
```

The way to reproduce the problem is just to restart kubelets on master nodes sequentially. Also note that just deleting kube apiserver pods doesn't seem to cause an issue.

The only interesting log that I could spot is:
```
coredns-7b678dd5bb-mnzzrcoredns-7b678dd5bb-2lf5w  corednscoredns E0111 15:00:26.752836       1 reflector.go:322] github.com/coredns/coredns/plugin/kubernetes/controller.go:322: Failed to watch *v1.Namespace: Get https://10.3.0.1:443/api/v1/namespaces?resourceVersion=140762277&timeoutSeconds=559&watch=true: dial tcp 10.3.0.1:443: connect: connection refused
 E0111 15:00:26.741853       1 reflector.go:322] github.com/coredns/coredns/plugin/kubernetes/controller.go:317: Failed to watch *v1.Endpoints: Get https://10.3.0.1:443/api/v1/endpoints?resourceVersion=140765670&timeoutSeconds=347&watch=true: dial tcp 10.3.0.1:443: connect: connection refused
ERROR: logging before flag.Parse: E0111 15:00:34.016793   12065 streamwatcher.go:109] Unable to decode an event from the watch stream: http2: server sent GOAWAY and closed the connection; LastStreamID=9, ErrCode=NO_ERROR, debug=""
```

During the time of failure we cannot resolve anything, `dig` gives:

```
dnstools# dig prometheus.sys-prom.svc.cluster.local 

; <<>> DiG 9.11.3 <<>> prometheus.sys-prom.svc.cluster.local
;; global options: +cmd
;; connection timed out; no servers could be reached
```

it looks like some kind of error that cannot be handled by the stream watcher and results in coredns failing to respond.

**Comment by chrisohaver** [comment]
> The way to reproduce the problem is just to restart kubelets on master nodes sequentially. Also note that just deleting kube apiserver pods doesn't seem to cause an issue.

Thanks.  I had tested restarting apiserver in the past to see how coredns handles API interruptions.
Interesting that restarting kubelet causes an api outage.

**Comment by chrisohaver** [comment]
> The way to reproduce the problem is just to restart kubelets on master nodes sequentially. 

@ffilippopoulos, just a kubelet restart is required?  Or do I need to create a new set of certs, then restart?

I'm having trouble replicating this issue doing just a kubelet restart.  Coredns continues to respond after the restart, no errors in log.

**Comment by ffilippopoulos** [comment]
@chrisohaver just restarting all master kubelets sequentially is enough for me to reproduce. 
In particular I run a `for` loop to `ssh` into kube masters and restart kubelets, while watching an `nslookup` to some in-cluster address from a `dnstools` pod.
 
I see some improvement by adding `readinessProbe`s on kube-apiservers but I cannot yet verify if that is enough to solve the issue.

**Comment by chrisohaver** [comment]
My cluster has a single master... perhaps that is a factor.  I'll set up a multi master cluster and try to replicate.

**Comment by chrisohaver** [comment]
Looks related: https://github.com/projectcalico/typha/issues/196

**Comment by mmack** [comment]
Yeah you're right... this looks indeed like the same problem... However i don't think coredns restarts itself if it looses the connection to the apiserver right?

**Comment by chrisohaver** [comment]
> ... ssh into kube masters and restart kubelets, while watching an nslookup to some in-cluster address from a dnstools pod.

For me, restarting kubelet interrupts connections to pods. E.g. if I restart kubelet, an interactive connection to a dnstools pod is broken, and cannot reconnect or spin up a new dnstools pod until kubelet is back up and running.

**Comment by chrisohaver** [comment]
>  However i don't think coredns restarts itself if it looses the connection to the apiserver right?

Correct, coredns isn't crashing -  it's related but not the same.

**Comment by ffilippopoulos** [comment]
> For me, restarting kubelet interrupts connections to pods. E.g. if I restart kubelet, an interactive connection to a dnstools pod is broken, and cannot reconnect or spin up a new dnstools pod until kubelet is back up and running.

I assume you have access to the nodes, so deploy `dnstools` and then ssh into the node that hosts it and docker exec into it. This is what I do to avoid what you described.

**Comment by johnbelamaric** [comment]
The typha issue looks like it's because of `glog` - we don't use that. If we lose connection to the api server we should be reconnecting - but we use standard client-go code for this, I believe

**Comment by george-angel** [comment]
@ffilippopoulos I think its important to note that restarting our kubelet on master also does an explicit `docker restart` of the api-server component:

- https://github.com/utilitywarehouse/tf_kube_ignition/blob/master/resources/master-kubelet.service#L27

**Comment by ffilippopoulos** [comment]
@mmack I can verify that in our case where we have multiple apiservers running adding a `readinessProbe` really helps and I am able to cycle master nodes without getting errors on `nslookup`s, but the issue should still stand in case of 1 master or when you lose connection to all of your apiservers at the same time.

**Comment by chrisohaver** [comment]
>  but we use standard client-go code for this, I believe

Yes.  Kubernetes' client-go package should IMO try to handle API restarts gracefully.  

Regarding methods for cycling certs, there are some solutions that allow for "self" cert cycling without restarts, but it seems those solutions are stalled in beta/alpha (since 1.8).

I still cannot reproduce this issue locally, so it's hard for me to identify precisely whats going on.  Here are a few possible causes...

1. One reason coredns could return a SERVFAIL is if any of the client-go api watches report that they are not synced to the local store (managed by client-go).  If this ever occurs, coredns reports SERVFAIL instead of NXDOMAIN when it processes a query for a record for which there is no match in the store.  However, coredns wont reports SERVFAIL when there _is_ a matching record in the store.  So, if this is the case, it could mean that the store is emptied/invalidated by client-go when this the API shuts down.  I couldn't find any client-go connection options that hint at controlling this kind of behavior.

2. In CoreDNS, we do create a custom IndexerInformer to trim down api objects before they are passed to the store. It's possible that this has introduced some issue.  I don't think this is the case, but we could test on a version prior to that change to confirm/eliminate the possibility.

3. Another possible culprit could be the full resync period we pass to client-go.  This defaults to 5 minutes.  The thinking here, that if a resync occurs while the api is down, perhaps the store is reset/emptied. But since a restart should take much less time than 5 minutes, and there is no mention of this issue being intermittent, I don't think this is the cause.  Easy to check this, by setting `resyncperiod` to 0 in the Corefile.

Until I can reproduce this locally I'm stabbing in the dark. It's possible that this only affects multi-master clusters (since all reporting the issue have multi-master clusters), so I'm currently trying to determine the quickest/easiest/cheapest way to build one.  I have not built a multi-master cluster before - any experienced advice is welcome.

**Comment by johnbelamaric** [comment]
I don't really get why restarting the kubelet would affect api server
access at all. Restarting the api server after rotating certs, if the
service account uses the old cert and it's not regenerated or rotated too,
could cause an issue. I haven't looked at this area closely before.

**Comment by mmack** [comment]
For us we have 2 or 3 replicas of coredns with 2 apiserver running. What 
about the others? How many replicas do you guys have? Does the error happen 
with only one replica? I think i can test that on monday...

Max

Am 18. Januar 2019 16:07:06 schrieb chrisohaver <notifications@github.com>:

**Comment by mmack** [comment]
ok, so here are my findings:

2 apiserver. 3 coredns instances. dnsutils running in one pod.

Go to the node where the pod was schedules. exec into it. Run:
```
while true; do dig +tries=1 +time=1 google.com; sleep 0.1; done
```

Then i go to the first master and do a systemctl restart apiserver.

Before:
```
; <<>> DiG 9.8.4-rpz2+rl005.12-P1 <<>> +tries=1 +time=1 google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 6143
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;google.com.			IN	A

;; ANSWER SECTION:
google.com.		62	IN	A	172.217.17.46

;; Query time: 0 msec
;; SERVER: 100.64.0.11#53(100.64.0.11)
;; WHEN: Mon Jan 21 10:17:33 2019
;; MSG SIZE  rcvd: 54

```

After restart:
```
; <<>> DiG 9.8.4-rpz2+rl005.12-P1 <<>> +tries=1 +time=1 google.com
;; global options: +cmd
;; connection timed out; no servers could be reached
```

To be clear: I did not change any certs. Just a plain apiserver restart.

Hope that helps...

**Comment by mmack** [comment]
Ah, another thing: If does not matter if i have 3 or just one coredns replica running. Same results.

**Comment by mmack** [comment]
Ok i digged in further: It does not happen with version 1.2.2. I happens above 1.3.0.

[Edit]
Got the same error now with version. 1.2.2. I'm out of ideas...

**Comment by chrisohaver** [comment]
>I don't really get why restarting the kubelet would affect api server
access at all. 

It would seem, at least in the terraform module below,  this is because kubelet's systemd unit file restarts it whenever kubelet is restarted ...

>  I think its important to note that restarting our kubelet on master also does an explicit docker restart of the api-server component:
> https://github.com/utilitywarehouse/tf_kube_ignition/blob/master/resources/master-kubelet.service#L27

**Comment by mmack** [comment]
```
docker logs f68579e98d40 -f
.:53
2019-01-23T06:22:56.539Z [INFO] CoreDNS-1.3.1
2019-01-23T06:22:56.539Z [INFO] linux/amd64, go1.11.4, 6b56a9c
CoreDNS-1.3.1
linux/amd64, go1.11.4, 6b56a9c
2019-01-23T06:22:56.539Z [INFO] plugin/reload: Running configuration MD5 = e1c7960649774ea6c3946b639aa6ee32





E0123 06:53:22.589260       1 streamwatcher.go:109] Unable to decode an event from the watch stream: http2: server sent GOAWAY and closed the connection; LastStreamID=33, ErrCode=NO_ERROR, debug=""
E0123 06:53:22.589260       1 streamwatcher.go:109] Unable to decode an event from the watch stream: http2: server sent GOAWAY and closed the connection; LastStreamID=33, ErrCode=NO_ERROR, debug=""
log: exiting because of error: log: cannot create log: open /tmp/coredns.coredns-5c6d4545b5-vgpqj.unknownuser.log.ERROR.20190123-065322.1: no such file or directory
```

fyi: I see this if i do a docker log directly on the host where coredns is running...

**Comment by mmack** [comment]
And finally i got it. A temp dir helps. Look at the typha issue above, same problem.

```
        volumeMounts:
        - name: config-volume
          mountPath: /etc/coredns
        - name: temp-volume
          mountPath: /tmp/
```

No i can restart my apiserver and don't see any dns queries getting lost...

**Comment by fischerman** [comment]
The problem is related to kubernetes/kubernetes#61006

I have a similar issue in case of network drops: 
```
W0122 19:58:48.174262       1 reflector.go:270] github.com/coredns/coredns/plugin/kubernetes/controller.go:315: watch of *v1.Service ended with: very short watch: github.com/coredns/coredns/plugin/kubernetes/controller.go:315: Unexpected watch close - watch lasted less than a second and no items received
log: exiting because of error: log: cannot create log: open /tmp/coredns.coredns-coredns-ccc9c4d9b-8djtp.unknownuser.log.WARNING.20190122-195848.1: no such file or directory
```

Basically, everytime the kubernetes client lib tries to log an error the container crashes. The same problem exists in the coredns helm chart. Can we add a /tmp folder to the Docker image as a workaround until the client libs are fixed?

**Comment by gjtempleton** [comment]
Yep, same problem as this: https://github.com/projectcalico/typha/issues/196 

Seems "best" practice when using `FROM scratch` is to also add a `/tmp` directory to allow glog to do its file writing thing.

**Comment by gjtempleton** [comment]
@mmack I've been doing a bit of playing around with this as well as raising #2525 

Have you definitely managed to reproduce this with versions of CoreDNS <1.3.1?

I've managed to successfully reproduce the behaviour with 1.3.1 but not _any_ version below, even post the switch to `FROM scratch` for the Dockerfile.

My suspicions are beginning to point towards https://github.com/coredns/coredns/commit/6b8c154441f4f4bcf5c8ee6cff888d28edac81fe given it happened between 1.3.0 and 1.3.1.

**Comment by mmack** [comment]
I'm pretty sure i've seen the errors below 1.3.1 however right now i'm just happy it's not happening anymore :)

**Comment by chrisohaver** [comment]
Closing, as I understand this to be fixed with #2529.  @mmack, Please reopen if this is still an issue.
